plan/predicate_push_down.go

```diff
@@ -138,6 +144,176 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
 	return
 }
 
+// outerJoinSimplifier simplify outer join.
+func outerJoinSimplify(p *Join, predicates []expression.Expression) error {
+	var innerTable, outerTable LogicalPlan
+	child1 := p.GetChildByIndex(0).(LogicalPlan)
+	child2 := p.GetChildByIndex(1).(LogicalPlan)
+	var fullConditions []expression.Expression
+	if p.JoinType == InnerJoin {
+		if leftChild, ok := child1.(*Join); ok {
+			fullConditions = concatOnAndWhereConds(p, predicates)
+			err := outerJoinSimplify(leftChild, fullConditions)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		if rightChild, ok := child2.(*Join); ok {
+			if fullConditions == nil {
+				fullConditions = concatOnAndWhereConds(p, predicates)
+			}
+			err := outerJoinSimplify(rightChild, fullConditions)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	} else if p.JoinType == LeftOuterJoin {
+		innerTable = child2
+		outerTable = child1
+	} else if p.JoinType == RightOuterJoin {
+		innerTable = child1
+		outerTable = child2
+	}
+	switch innerPlan := innerTable.(type) {
+	case *DataSource:
+		canBeSimplified := false
+		for _, expr := range predicates {
+			if canBeSimplified {
+				break
+			}
+			switch x := expr.(type) {
+			case *expression.Constant:
+				if x.Value.IsNull() {
+					canBeSimplified = true
+				} else if isTrue, err := x.Value.ToBool(); err != nil || isTrue == 0 {
+					if err != nil {
+						return errors.Trace(err)
+					}
+					canBeSimplified = true
+				}
+			case *expression.ScalarFunction:
+				isOk, err := isNullRejected(innerPlan.GetSchema(), x)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if isOk {
+					canBeSimplified = true
+				}
+			}
+		}
+		if canBeSimplified {
+			p.JoinType = InnerJoin
+			if _, ok := outerTable.(*Join); ok {
+				fullConditions = concatOnAndWhereConds(p, predicates)
+				err := outerJoinSimplify(outerTable.(*Join), fullConditions)
+				if err != nil {
+					return errors.Trace(err)
+				}
+			}
+		} else {
+			return nil
+		}
+	case *Join:
+		fullConditions = concatOnAndWhereConds(p, predicates)
+		err := outerJoinSimplify(innerPlan, fullConditions)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if innerPlan.JoinType != InnerJoin {
+			break
+		}
+		if x, ok := outerTable.(*Join); ok {
+			fullConditions = concatOnAndWhereConds(innerPlan, fullConditions)
+			err := outerJoinSimplify(x, fullConditions)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+// isNullRejected check whether a condition is null-rejected
+// A condition would be null-rejected in one of following cases:
+// If it is a predicate containing a reference to an inner table that evaluates to UNKNOWN or FALSE when one of its arguments is NULL.
+// If it is a conjunction containing a null-rejected condition as a conjunct.
+// If it is a disjunction of null-rejected conditions.
+func isNullRejected(schema expression.Schema, scalarFunc *expression.ScalarFunction) (bool, error) {
```

Comment:
you should combine isNullRejected and calculateResultOfScalarFunc to one function, return nil that stands for unknown result.